substrate/frame/revive/src/vm/evm/instructions/host.rs

```diff
@@ -15,251 +15,220 @@
-	let Some(code) = context.host.load_account_code(address) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
-	let spec_id = context.interpreter.runtime_flag.spec_id();
-	if spec_id.is_enabled_in(BERLIN) {
-		gas_legacy!(context.interpreter, warm_cold_cost(code.is_cold));
-	} else if spec_id.is_enabled_in(TANGERINE) {
-		gas_legacy!(context.interpreter, 700);
-	} else {
-		gas_legacy!(context.interpreter, 20);
-	}
-
-	*top = U256::from(code.len());
+	gas!(context.interpreter, RuntimeCosts::CodeSize);
+	let h160 = sp_core::H160::from_slice(&top.to_be_bytes::<32>()[12..]);
+	let code_size = context.interpreter.extend.code_size(&h160);
+	*top = U256::from(code_size);
 }
 
 /// EIP-1052: EXTCODEHASH opcode
 pub fn extcodehash<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-	check!(context.interpreter, CONSTANTINOPLE);
 	popn_top!([], top, context.interpreter);
-	let address = top.into_address();
-	let Some(code_hash) = context.host.load_account_code_hash(address) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
-	let spec_id = context.interpreter.runtime_flag.spec_id();
-	if spec_id.is_enabled_in(BERLIN) {
-		gas_legacy!(context.interpreter, warm_cold_cost(code_hash.is_cold));
-	} else if spec_id.is_enabled_in(ISTANBUL) {
-		gas_legacy!(context.interpreter, 700);
-	} else {
-		gas_legacy!(context.interpreter, 400);
-	}
-	*top = code_hash.into_u256();
+	gas!(context.interpreter, RuntimeCosts::CodeHash);
+	let h160 = sp_core::H160::from_slice(&top.to_be_bytes::<32>()[12..]);
+	let code_hash = context.interpreter.extend.code_hash(&h160);
+	*top = U256::from_be_bytes(code_hash.0);
 }
 
 /// Implements the EXTCODECOPY instruction.
 ///
 /// Copies a portion of an account's code to memory.
 pub fn extcodecopy<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 	popn!([address, memory_offset, code_offset, len_u256], context.interpreter);
-	let address = address.into_address();
-	let Some(code) = context.host.load_account_code(address) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
-
 	let len = as_usize_or_fail!(context.interpreter, len_u256);
-	gas_or_fail_legacy!(
-		context.interpreter,
-		gas::extcodecopy_cost(context.interpreter.runtime_flag.spec_id(), len, code.is_cold)
-	);
+
+	gas!(context.interpreter, RuntimeCosts::ExtCodeCopy(len as u32));
+	let address = sp_core::H160::from_slice(&address.to_be_bytes::<32>()[12..]);
+
 	if len == 0 {
 		return;
 	}
 	let memory_offset = as_usize_or_fail!(context.interpreter, memory_offset);
-	let code_offset = min(as_usize_saturated!(code_offset), code.len());
+	let code_offset = as_usize_saturated!(code_offset);
+
 	resize_memory!(context.interpreter, memory_offset, len);
 
-	// Note: This can't panic because we resized memory to fit.
-	context.interpreter.memory.set_data(memory_offset, code_offset, len, &code);
+	let mut buf = context.interpreter.memory.slice_mut(memory_offset, len);
```

Comment:
I think we should keep the comment. Cause `slice_mut` would probably panic of out of bounds.